ISPN-13616 Stream pipeline blocks if filterKeys() gets an empty set

wburns opened the pull request:
https://issues.redhat.com/browse/ISPN-13616
https://issues.redhat.com/browse/ISPN-13628

wburns commented:
Replaces #9785

heli80 commented on core/src/main/java/org/infinispan/reactive/publisher/impl/ClusterPublisherManagerImpl.java:
```diff
@@ -223,12 +227,19 @@ public void stop() {
 
+      int keyTargetSize = keyTargets.size();
+
+      if (keyTargetSize == 0) {
+         flowableProcessor.onComplete();
+         return;
+      }
```
Couldn't the context contain entries that need to be processed, while `keyTargetSize` is 0?

wburns replied:
Good point, I will take a look again. 👍

wburns replied:
This is fine for the keyPublisher as it checks the context, but the entryPublisher does not. I will add that in.

heli80 replied:
Cool :+1:

heli80 commented on core/src/main/java/org/infinispan/reactive/publisher/impl/ClusterPublisherManagerImpl.java:
```diff
@@ -223,12 +227,19 @@ public void stop() {
+      int keyTargetSize = keyTargets.size();
+
+      if (keyTargetSize == 0) {
+         flowableProcessor.onComplete();
```
I think just calling `onComplete()` is maybe not enough. Because, if the `transformer` function is not part of the (empty) pipeline then (for some cases) no "identity" element is returned finally.

The cases where this will happen is when this method of `DistributedCacheStream` is used:
```
   public R reduce(R identity, BinaryOperator<R> accumulator) {
      return performPublisherOperation(PublisherReducers.reduce(identity, accumulator),
            PublisherReducers.reduce(accumulator));
   }
```
Here only the `transformer` but not the `finalizer` takes care of the "identity" element.
I can create a unit-test to show the point if you want.

wburns replied:
I understand what you are saying and it is a good point, This is important for other Flowable mappings such as `Flowable.switchIfEmpty` etc. I was being a bit overzealous on the code reductions and your prior PR is probably the one we will want to use.

wburns replied:
But I may have something else still. Either way I will a couple more test cases, unless you have one available on you I can copy ;)

heli80 replied:
Sure, what do you think about this testcase?
```
   @Test(dataProvider = "GuaranteeParallelEntry")
   public void testReduceTransformerOnlyIdentityEmptyKeys(DeliveryGuarantee deliveryGuarantee, boolean parallel, boolean isEntry) {
      Cache<Integer, String> cache = cache(0);
      int insertAmount = insert(cache).size();
      assertTrue(insertAmount > 0);

      // This empty set filters everything, but the computation must still complete
      Set<Integer> keysToInclude = new HashSet<>();

      // The identity element must still be returned, even if only the transformer (but not the finalizer) function takes care of it
      ClusterPublisherManager<Integer, String> cpm = cpm(cache);
      Object identity;
      CompletionStage<?> stage;
      if (isEntry) {
         SerializableBinaryOperator<CacheEntry<Integer, String>> binOp = (e1, e2) -> e1;
         identity = NullCacheEntry.getInstance();
         stage = cpm.entryReduction(parallel, null, keysToInclude, null, false, deliveryGuarantee,
               PublisherReducers.reduce((CacheEntry<Integer, String>) identity, binOp), PublisherReducers.reduce(binOp));
      } else {
         SerializableBinaryOperator<Integer> binOp = (k1, k2) -> k1;
         identity = 0;
         stage = cpm.keyReduction(parallel, null, keysToInclude, null, false, deliveryGuarantee,
               PublisherReducers.reduce((Integer) identity, binOp), PublisherReducers.reduce(binOp));
      }

      Object actual = stage.toCompletableFuture().join();
      Object expected = identity;
      assertTrue(expected == actual);
   }
```

wburns replied:
Okay, I was able to fix the reduction issue. However, looking at Publisher there is a different issue that I am not sure if I can resolve or not. I will update later though.

danberindei commented on core/src/test/java/org/infinispan/reactive/publisher/impl/SimpleClusterPublisherManagerTest.java:
```diff
@@ -172,6 +173,39 @@ public void testCountSpecificKeys(DeliveryGuarantee deliveryGuarantee, boolean p
+   @Test(dataProvider = "GuaranteeParallelEntry")
+   public void testCountEmptyKeys(DeliveryGuarantee deliveryGuarantee, boolean parallel, boolean isEntry) {
+      Cache<Integer, String> cache = cache(0);
+      int insertAmount = insert(cache).size();
```
Nitpicking: Inserting 24 entries seems excessive in all the tests here, but even more when the test is not reading any of the entries.

wburns replied:
The test requires something to confirm it actually works properly. But I can just reduce all the tests from 24 to 12 for example maybe. TBH it isn't going to make that much of an actual difference. In prior examples we were inserting tens of thousands, which was obviously an issue.

heli80 commented on core/src/main/java/org/infinispan/reactive/publisher/impl/ClusterPublisherManagerImpl.java:
```diff
@@ -217,6 +217,28 @@ public void stop() {
+      // Need to do this in case if the transformer or finalizer has additional values such as reduce with identity
+      // or switchIfEmpty etc.
+      CompletionStage<R> transformedStage = transformer.apply(Flowable.empty());
+      if (finalizer != null) {
+         transformedStage = transformedStage.thenApply(value -> finalizer.apply(Flowable.just(value)))
+               .thenCompose(Function.identity());
+      }
```
I'm not really sure why the application of `finalizer` is needed here. My original PR war aligned to the implementation of `handleContextInvocation` and there finalizer is also not used.

Maybe the finalizer anyway applies in the end since this line is in place in `ClusterPublisherManagerImpl.reduction()`?
```
      // Apply the finalizer first (which subscribes) before emitting items, to avoid buffering in UnicastProcessor
      CompletionStage<R> stage = finalizer.apply(flowableProcessor);
```

wburns replied:
Finalizer is designed to be allowed to run as many times as you have required subset of entries. In this case, it is extraneous for the call as you pointed out.

To describe the use of finalizer more, it is used in the place you mentioned to consolidate responses from each of the nodes into a single value. However, the finalizer must also be used on each of the remote/local nodes if the reduction is parallel or we have more strict GUARANTEEs which are done per segment. But since we have a single value, it isn't really needed at all tbh :)

heli80 replied:
Thanks for the explanation! Reactive code really drives me crazy sometimes ;-)

heli80 commented on core/src/main/java/org/infinispan/reactive/publisher/impl/ClusterPublisherManagerImpl.java:
```diff
@@ -217,6 +217,28 @@ public void stop() {
+      // Need to do this in case if the transformer or finalizer has additional values such as reduce with identity
+      // or switchIfEmpty etc.
+      CompletionStage<R> transformedStage = transformer.apply(Flowable.empty());
+      if (finalizer != null) {
+         transformedStage = transformedStage.thenApply(value -> finalizer.apply(Flowable.just(value)))
+               .thenCompose(Function.identity());
+      }
```
Another issue is that `transformedStage` could complete with `null` and `Flowable.just` would then throw an NPE. This happens when `DistributedCacheStream.reduce(BinaryOperator<R>)` (specifies no identity element) is used.

I crafted this unit-test to show the point:
```
   @Test(dataProvider = "GuaranteeParallelEntry")
   public void testReduceWithoutIdentityEmptyKeys(DeliveryGuarantee deliveryGuarantee, boolean parallel, boolean isEntry) {
      Cache<Integer, String> cache = cache(0);
      int insertAmount = insert(cache).size();
      assertTrue(insertAmount > 0);

      // This empty set filters everything, but the computation must still complete
      Set<Integer> keysToInclude = new HashSet<>();

      // Using a reduce function without explicit identity tests the special case where completion-stages terminate with null
      ClusterPublisherManager<Integer, String> cpm = cpm(cache);
      CompletionStage<?> stage;
      if (isEntry) {
         Function<Publisher<CacheEntry<Integer, String>>, CompletionStage<CacheEntry<Integer, String>>> function =
               PublisherReducers.reduce((SerializableBinaryOperator<CacheEntry<Integer, String>>) (e1, e2) -> e1); // reduce without identity
         stage = cpm.entryReduction(parallel, null, keysToInclude, null, false, deliveryGuarantee,
               function, function);
      } else {
         Function<Publisher<Integer>, CompletionStage<Integer>> function =
               PublisherReducers.reduce((SerializableBinaryOperator<Integer>) (k1, k2) -> k1); // reduce without identity
         stage = cpm.keyReduction(parallel, null, keysToInclude, null, false, deliveryGuarantee,
               function, function);
      }

      Object result = stage.toCompletableFuture().join();
      assertTrue(result == null);
   }
```

**NOTE:** This unit-test comes from another (really small) PR I was preparing, so it would also fail with my original PR code. Wanted to see how this PR goes till I submit the next ;-)

wburns replied:
This is fixed by the previous comment, just removing the finalizer as the whenComplete properly handles null. But I can add this test as well.

wburns commented:
Updated with some more tests and added some documentation around the publisher methods to talk about using methods that may add values per Publisher versus by entry. Also added some changes for handling exceptions when submitting the commands such as marshalling.

heli80 commented:
Is there anything I can do or help to get this PR finally finished?
> This is fine for the keyPublisher as it checks the context, but the entryPublisher does not. I will add that in.

I guess this is still open...

wburns commented:
Actually the entryPublisher is currently done at a different point so it is fine. This PR is just waiting for someone else to review is all :)

pruivo reviewed (commented):
LGTM, just minor comments/questions :)

wburns commented on core/src/main/java/org/infinispan/reactive/publisher/impl/ClusterPublisherManager.java:
```diff
@@ -34,6 +35,14 @@
     * are then fed into the <b>finalizer</b> to produce a final result. If publisher is parallel the <b>finalizer</b>
     * will be invoked on each node to ensure there is only a single result per node.
+    * <p>
+    * The provided <b>transformer</b> must be able to be run multiple times on any number of sub groups of elements in the cache.
```
Alright.

wburns replied:
Removed this and duplicated below.

wburns commented on core/src/main/java/org/infinispan/reactive/publisher/impl/ClusterPublisherManager.java:
```diff
@@ -34,6 +35,14 @@
+    * The provided <b>transformer</b> must be able to be run multiple times on any number of sub groups of elements in the cache.
+    * This is because we may need to perform the transformation on each segment within the cache which hold distinct data.
+    * <p>
+    * Due to the reduction nature of this method it is possible to add arbitrary values as a value to returned
```
I tried rewording a bit, let me know what you think.

wburns commented:
Addressed comments and rebased.

pruivo commented:
merged! thanks @wburns !
